heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -609,6 +610,63 @@ data SimplImpl ps_in ps_out where
     ExprVar (LLVMPointerType w) -> LLVMFieldPerm w sz ->
     SimplImpl (RNil :> LLVMPointerType w) (RNil :> LLVMPointerType w)
 
+  -- | Split an LLVM field permission that points to a known bitvector:
+  --
+  -- > x:[l]ptr((rw,off) |-> eq(bv1++bv2))
+  -- >   -o [l]x:ptr((rw,off) |-> eq(bv1))
+  -- >      * [l]x:ptr((rw,off+len(bv1)) |-> eq(bv2))
+  --
+  -- Note that the definition of @++@ depends on the current endianness.
+  SImpl_SplitLLVMWordField ::
+    (1 <= w, KnownNat w, 1 <= sz1, KnownNat sz1, 1 <= sz2, KnownNat sz2,
+     1 <= (sz2 - sz1), KnownNat (sz2 - sz1)) =>
```

Comment:
Do we really need `KnownNat (sz2 - sz1)` or do we just need a `NatRepr (sz2 - sz1)`? Because the latter can be gotten from [`subNat`](https://hackage.haskell.org/package/parameterized-utils-2.1.4.0/docs/Data-Parameterized-NatRepr.html#v:subNat). Same goes for the other `KnownNat` instances of `+` and `-` below.

Reply:
All of these instances are available at the point in the implication prover where the constructor is applied, so it makes sense to just save them rather than re-generating them in `simplImplIn` and `simplImplOut`. It just makes the code simpler in those functions.